modules/docs/arrow-docs/docs/docs/arrow/typeclasses/traverse/README.md

```diff
@@ -11,16 +11,366 @@ redirect_from:
+data class SecurityError(val message: String)
+data class Credentials(val email: String, val password: String)
+
+fun validateLogin(c: Credentials): Either<SecurityError, Unit> = TODO()
+
+fun userInfo(u: User): Promise<ForIO, Profile> = TODO()
+
+data class Admin(val name: String, val crd: Credentials): User
+
+fun ListK<User>.processLogin()
+```
+
+Evidently, `Traverse` is not limited to `List` or `Nel`, it provides an abstraction over 'things that can be traversed over', like a Binary tree, hence the name `Traverse`.
+
+#### Playing with `Reader`
+
+Another interesting effect we can use is Reader. Recall that a `Reader<D, A>` is a type alias for `Kleisli<ForId, D, A>` which is a wrapper around `(D) -> A`.
+
+If we fix `D` to be some sort of dependency or configuration, we can use the `Reader` applicative in our `traverse`.
+
+```kotlin:ank
+import arrow.mtl.Reader
+
+interface Context
+interface Topic
+interface Result
+
+typealias Job<A> = Reader<Context, A>
+
+fun processTopic(t: Topic): Job<Result> = TODO()
+```
+
+We can imagine we have a data pipeline that processes a bunch of data, each piece of data being categorized by a topic. Given a specific topic, we produce a `Job` that processes that topic. (Note that since a `Job` is just a `Reader`/`Kleisli`, one could write many small `Jobs` and compose them together into one `Job` that is used/returned by `processTopic`.)
+
+Corresponding to our bunches of data are bunches of topics, a `List<Topic>` if you will. Since Reader has an `Applicative` instance, we can `traverse` over this list with `processTopic`.
+
+```kotlin:ank
+import arrow.core.Id
+import arrow.core.ForId
+import arrow.core.ListK
+import arrow.core.extensions.id.applicative.applicative
+import arrow.mtl.KleisliPartialOf
+import arrow.mtl.extensions.kleisli.applicative.applicative
+import arrow.mtl.fix
+
+fun processTopics(topics: ListK<Topic>): Job<ListK<Result>> =
+  topics.traverse<KleisliPartialOf<ForId, Context>, Result>(Job.applicative<ForId, Context>(Id.applicative())) {
+    processTopic(it)
+  }.fix()
+```
+
+Note the nice return type - `Job<List<Result>>`. We now have one aggregate `Job` that when run, will go through each topic and run the topic-specific job, collecting results as it goes. We say "when run" because a `Job` is some function that requires a Context before producing the value we want.
+
+One example of a "context" can be found in the [Spark](http://spark.apache.org/) project. In Spark, information needed to run a Spark job (where the master node is, memory allocated, etc.) resides in a `SparkContext`. Going back to the above example, we can see how one may define topic-specific Spark jobs `(type Job<A> = Reader<SparkContext, A>)` and then run several Spark jobs on a collection of topics via `traverse`. We then get back a `Job<List<Result>>`, which is equivalent to (`SparkContext) -> List<Result>`. When finally passed a `SparkContext`, we can run the job and get our results back.
```

Comment:
I'd also add an example of Android's Context.

Reply:
I purpose to add a more tagless final approach, so Users are not inclined to use MTL.
What do you think @pakoito ?